app/models/access.rb

```diff
@@ -0,0 +1,38 @@
+class Access < ApplicationRecord
+  ALLOWED_RESOURCE_TYPES = %w[Organization FacilityGroup Facility].freeze
+
+  belongs_to :user, required: true
+  belongs_to :resource, polymorphic: true, optional: true
+
+  enum role: {
+    super_admin: "super_admin",
+    admin: "admin",
+    analyst: "analyst"
+  }
+
+  validates :role, presence: true
+  validates :resource_type, inclusion: {in: ALLOWED_RESOURCE_TYPES}, allow_nil: true
+
+  class << self
+    def organizations
+      resources_for(Organization)
+    end
+
+    def facility_groups
+      resources_for(FacilityGroup)
+        .or(FacilityGroup.where(organization: resources_for(Organization)))
+    end
+
+    def facilities
+      resources_for(Facility)
+        .or(Facility.where(facility_group: FacilityGroup.where(organization: resources_for(Organization))))
+        .or(Facility.where(facility_group: resources_for(FacilityGroup)))
```

Comment:
As much as I love ActiveRecord, for a query as crucial as this I think its often preferable to drop into SQL to make sure its doing what you want. Finding the right facilities is critical to our whole authorization system, and bugs here could have really negative impacts. Are you sure this kind of `or` chaining does the right thing if there are multiple hits along the chain?  For example, if a user has read access to a FacilityGroup and then write access to a Facility inside that group.

Compare this to a [SQL approach](https://github.com/simpledotorg/simple-server/pull/1173/files#diff-a9804da7022a39203c40f4586796c96fR31), where there are more lines of code but it is super clear what is going on:

```
SELECT facilities.* 
FROM facilities 
  INNER JOIN facility_groups ON facility_groups.id = facilities.facility_group_id 
  INNER JOIN organizations ON facility_groups.organization_id = organizations.id 
  INNER JOIN roles ON roles.resource_type = 'Organization' 
  AND roles.resource_id = organizations.id 
WHERE 
  roles.user_id = : user_id 
  AND roles.name in (: roles) 

UNION 

SELECT facilities.* 
FROM facilities 
  INNER JOIN facility_groups ON facility_groups.id = facilities.facility_group_id 
  INNER JOIN roles ON roles.resource_type = 'FacilityGroup' 
  AND facility_groups.id = roles.resource_id 
WHERE 
  roles.user_id = : user_id 
  AND roles.name in (: roles) 

UNION 

SELECT facilities.* 
FROM facilities 
  INNER JOIN roles ON roles.resource_type = 'Facility' 
  AND roles.resource_id = facilities.id 
WHERE 
  roles.user_id = : user_id 
  AND roles.name in (: roles)
```

Reply:
I disagree pretty strongly.

> Are you sure this kind of or chaining does the right thing

Yes, because we tested it and you can simply look at the SQL query to confirm it's true. Just because you write SQL does not mean your query is correct. 

For example, the query you pasted could very well be incorrect because you will end up querying potentially discarded facilities, organizations or facility groups.

If we don't trust our own abstractions, we shouldn't use them. Adding nuances like this actually makes *me* nervous.

Separately, I think this query can be **even** further simplified, take a look at [Take 2](https://github.com/simpledotorg/simple-server/pull/1160/files#diff-dbee6fa62dba5eb76fea1e205e2a5f54R52).

Reply:
I would also lean towards leveraging ActiveRecord to perform these queries ☝️ Whether we use AR or SQL, we're gonna have to unit test this thing to the moon for maximum confidence. I don't think _how_ we implement makes it easier/harder to test, and AR has some advantages as Kitty mentions.

Reply:
I agree that however we write this, it should be tested like crazy. 👍 

I can go either way on how this is written, my larger concern is that the external API for permissions is as simple as possible and doesn't require a developer to come looking at the underlying implementation to figure out what is going.